docs/proposals/shoot-control-plane-migration/README.md

```diff
@@ -0,0 +1,234 @@
+   
+4. A ring of highly available distributed cache among the seed clusters.
+
+   This approach has the potential benefit of avoiding the access to garden cluster if the distributed cache can be actively kept up-to-date by gardener.
+
+###### Lease data structure for a shoot
+
+* Owning seed cluster
+* Lease Period
+  * Lease start timestamp
+  * Lease end timestamp or interval
+
+##### Shoot Control-plane Co-ordinator
+
+The Shoot Control-plane Co-ordinator is a pod (actually, a deployment) deployed as a part of each of the shoot control-plane in the seed clusters.
+
+It has the responsibility to monitor whether the current seed cluster where it is deployed has the lease for the given shoot using any of the mechanisms chosen [above](#advertising-current-seed-cluster-lease-for-a-shoot-cluster) and take appropriate action to activate/passivate the relevant shoot control-plane components such as machine-controller-manager, etcd statefulset, VPN etc.
+
+It has two states.
+
+1. Active
+
+In the active state, the co-ordinator is able to reach the lease service and knows that the current seed is the owner for the given shoot according to the lease service.
+
+In such a scenario, the co-ordinator makes sure the shoot control-plane components such as machine-controller-manager, etcd statefulset, VPN etc. are scaled up and active.
+
+2. Passive
+
+In the passive state, the co-ordinator is either not able reach the lease service or knows that the current seed is not the owner of the given shoot according to the lease service.
+
+In such a scenario, the co-ordinator passivates the shoot control-plane components such as machine-controller-manager, etcd statefulset, VPN etc. by scaling them down to zero.
+
+It should not scale itself down, of course.
+
+##### State Diagram
+
+![State Diagram](state-diagram.png)
+
+This component should also expose a `/healthz` endpoint which returns healthy only if it can get the current lease status and if the current seed cluster is the owner of the lease for the given shoot. This `/healthz` endoint can be used for livenessProbe in the shoot machine-controller-manager, shoot etcd and also perhaps in the VPN components to avoid co-ordination problems in case the shoot control-plane co-ordinator itself crashes.
+
+###### Melt-down scenario
+
+In the [above](#shoot-control-plane-co-ordinator) approach, there is a possibility of a melt-down-like scenario where the co-ordinator is not able to access the lease status for some long period even though it actually owns the lease during that period, but the gardener is able to access the seed cluster. In such a scenario, co-ordinator will scale down the relevant component to zero while a gardener reconciliation will scale them back up to the original values. This cycle can potentially keep happening while the scenario persists.
+
```

Comment:
Well, the coordinator is fine. But if the DNS approach works, you wouldn't even have to worry about melt-down in the first place (if it happens, everything breaks down, but DNS is the most battle-proven service there is in the cloud business with decades of DoS and ops experience behind it - it usually never fails and if it does, no action must be taken anyhow because of the other consequences).

Plus, the DNS record is something that the Gardener maintains in its flow logic (the actual change runs in the seed cluster though, i.e. this behaviour might need some adaptation to make it safe also in corner cases, e.g. when the configuration is expressed and the job started, but then the seed fails and comes back up again, but that's a theoretical corner case).

Reply:
With DNS, it is not clear to me how to handle the case where the source seed is out of reach of gardener. I need to check more.

Reply:
The lease is basically the DNS A record from the API FQDN to the LB IP (or CNAME if the LB has an A record already). The Gardener can change it in Route53, whether or not the API server is reachable. If it's changed and the seed sees it has fallen out of fashion, it can shut down.

Reply:
What if the seed does not see the DNS record change for some time due to some network issues?

Reply:
DNS is HA. If it can't resolve the API server, it's most likely unavailable for everybody. More dangerous is the caching/TTL. One would need to think that through in more detail, but DNS has its merits.

Reply:
Agreed about the potential of DNA as compared with the other approaches for a lease server. Just trying to see how some of the gaps can be filled.

Reply:
I believe, DNS would work fine, TTL 60s. If a control plane coordinator comes up, it checks the DNS record. From the first time it sees itself as active, it waits 60+s before it commences. By that time, any other control plane should have noticed, it lost the DNS record/lease. If there was any issue and couldn't check the record, local implementation could ensure/freeze all logic. Meaning: the coordinator only or all components if there is none, could freeze when it sees that the last time it checked the DNS record was more than 60s ago.

Reply:
Yes. DNS looks promising. We will do a small evaluation to make sure it works as expected.